Add soft quadrupole

cemitch99 opened the pull request:
Implementation of quadrupole w/ soft-edge fringe fields.

Math for map computation, default coefficients, and example parameters have been updated.

- [x] Modify composite maps used for numerical integration
- [x] Modify default on-axis coefficients
- [x] Modify example inputs & docs

cemitch99 committed:
Examples for 3D space charge benchmarking

- Modified the initial beam size in the IOTA lens benchmark example.
- Added 2 benchmarks of 3D space charge for initial testing.
- Add documentation for 2 benchmarks with space charge.
- Add a benchmark example with space charge and periodic s-dependent focusing.
- Added an s-dependent example using a Kurth beam without space charge.
- Modified tolerance for IOTA lens benchmark example.
  Reduced tolerance to account for smaller initial beam size and
  improved preservation of invariants of motion.
- Modified tolerances of space charge examples to allow CI tests to
  pass when space charge is not active.

- Modified tolerance for space charge examples.
  These should fail unless space charge is turned on.

cemitch99 committed:
Update input_kurth_10nC.in

Selected numerical values for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Added soft-edge quadrupole; preliminary pattern only.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Delete input_kurth_10nC.in

This is not part of this pull request...

ax3l commented on examples/quadrupole_softedge/README.rst:
```diff
@@ -0,0 +1,54 @@
+.. _examples-solenoid-softedge:
+
+Soft-edge solenoid
+===================
```
```suggestion
Soft-Edge Solenoid
==================
```

ax3l commented on examples/quadrupole_softedge/README.rst:
```diff
@@ -0,0 +1,54 @@
+Run
+---
+
+This example can be run as a Python script (``python3 run_solenoid_softedge.py``) or with an app with an input file (``impactx input_solenoid_softedge.in``).
```
```suggestion
This example can be run as a Python script (``python3 run_quadrupole_softedge.py``) or with an app with an input file (``impactx input_quadrupole_softedge.in``).
```

ax3l commented on examples/quadrupole_softedge/README.rst:
```diff
@@ -0,0 +1,54 @@
+
+   .. tab-item:: Python Script
+
+       .. literalinclude:: run_solenoid_softedge.py
+          :language: python3
+          :caption: You can copy this file from ``examples/solenoid_softedge/run_solenoid_softedge.py``.
+
+   .. tab-item:: App Input File
+
+       .. literalinclude:: input_solenoid_softedge.in
+          :language: ini
+          :caption: You can copy this file from ``examples/solenoid_softedge/input_solenoid_softedge.in``.
```
```suggestion
       .. literalinclude:: run_quadrupole_softedge.py
          :language: python3
          :caption: You can copy this file from ``examples/quadrupole_softedge/run_quadrupole_softedge.py``.

   .. tab-item:: App Input File

       .. literalinclude:: input_quadrupole_softedge.in
          :language: ini
          :caption: You can copy this file from ``examples/quadrupole_softedge/input_quadrupole_softedge.in``.
```

ax3l commented on examples/quadrupole_softedge/README.rst:
```diff
@@ -0,0 +1,54 @@
+
+We run the following script to analyze correctness:
+
+.. dropdown:: Script ``analysis_solenoid_softedge.py``
+
+   .. literalinclude:: analysis_solenoid_softedge.py
+      :language: python3
+      :caption: You can copy this file from ``examples/solenoid_softedge/analysis_solenoid_softedge.py``.
```
```suggestion
.. dropdown:: Script ``analysis_quadrupole_softedge.py``

   .. literalinclude:: analysis_quadrupole_softedge.py
      :language: python3
      :caption: You can copy this file from ``examples/quadrupole_softedge/analysis_quadrupole_softedge.py``.
```

ax3l commented on examples/quadrupole_softedge/input_quadrupole_softedge.in:
```diff
@@ -0,0 +1,42 @@
+quad1.ds = 6.0
+quad1.gscale = 1.233482899483985
+quad1.mapsteps = 800
+
```
```suggestion


```

ax3l commented on src/python/elements.cpp:
```diff
@@ -245,4 +245,21 @@ void init_elements(py::module& m)
+
+    py::class_<SoftQuadrupole, elements::Thick>(me, "SoftQuadrupole")
+        .def(py::init<
+                 amrex::ParticleReal const,
+                 amrex::ParticleReal const,
+                 std::vector<amrex::ParticleReal>,
+                 std::vector<amrex::ParticleReal>,
+         int const,
+                 int const
```
```suggestion
                 amrex::ParticleReal,
                 amrex::ParticleReal,
                 std::vector<amrex::ParticleReal>,
                 std::vector<amrex::ParticleReal>,
                 int,
                 int
```

cemitch99 committed:
Update README.rst

Updated example README documentation.

cemitch99 committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Modify math for soft quad map push.

cemitch99 committed:
Update analysis_quadrupole_softedge.py

Increase tolerance artificially until coefficients and example are updated, to aid testing.

cemitch99 committed:
Update analysis_quadrupole_softedge.py

Relax tolerance further.

cemitch99 committed:
Add a working benchmark example for soft-edge quadrupole.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented:
All tests now pass for the soft quadrupole except CUDA / NVCC 11.0.2 SP.

cemitch99 committed:
Update README.rst

Updated benchmark test documentation.

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -225,6 +225,22 @@ Lattice Elements
+            * ``<element_name>.gscale`` (``float``, in inverse meters) Scaling factor for on-axis magnetic field gradient
+
+            * ``<element_name>.cos_coefficients`` (array of ``float``) cos coefficients in Fourier expansion of the on-axis field gradient
+              (optional); default is a thin-shell model from `DOI:10.1016/J.NIMA.2022.166706 <https://doi.org/10.1016/j.nima.2022.166706>`__
```
Discussed today: we are looking for a good reference we can use for a soft edge quad as data input.

If we don't find a good experimental model, we could use model coefficients after the hyberbolic tangent model used in MaryLie.

cemitch99 replied:
Done.

cemitch99 committed:
Modify default coefficients of on-axis quadrupole field gradient.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update parameters.rst

Updated soft-edge quad documentation of cos_coef and sin_coef to include default values.

cemitch99 committed:
Update python.rst

Update Python documentation of default cos_coef and sin_coef values.

ax3l committed:
Remove leftover debug print

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -225,6 +225,22 @@ Lattice Elements
+              (optional); default is a tanh fringe field model from `<http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
+
+            * ``<element_name>.sin_coefficients`` (array of ``float``) sin coefficients in Fourier expansion of the on-axis field gradient
+              (optional); default is a tanh fringe field model from `<http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
```
```suggestion
              (optional); default is a tanh fringe field model from `MaryLie 3.0 <http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
```

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -225,6 +225,22 @@ Lattice Elements
+            * ``<element_name>.gscale`` (``float``, in inverse meters) Scaling factor for on-axis magnetic field gradient
+
+            * ``<element_name>.cos_coefficients`` (array of ``float``) cos coefficients in Fourier expansion of the on-axis field gradient
+              (optional); default is a tanh fringe field model from `<http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
```
Title of the link missing:
```suggestion
              (optional); default is a tanh fringe field model from `MaryLie 3.0 <http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
```

ax3l commented on examples/quadrupole_softedge/README.rst:
```diff
@@ -0,0 +1,46 @@
+.. _examples-quadrupole-softedge:
+
+Soft-Edge Quadrupole
+===================
```
```suggestion
Soft-Edge Quadrupole
====================
```

ax3l commented on examples/quadrupole_softedge/README.rst:
```diff
@@ -0,0 +1,46 @@
+Soft-Edge Quadrupole
+===================
+
+This is a modification of the test for a matched electron beam propagating through a stable
+FODO cell, in which the quadrupoles have been replaced with soft-edge quadrupole
```
Let's link the test we mean. You refer to `examples/fodo/`, right? :)
```suggestion
This is a modification of :ref:`the test for a matched electron beam propagating through a stable FODO cell <examples-fodo>`,
in which the quadrupoles have been replaced with soft-edge quadrupole
```

ax3l commented on examples/quadrupole_softedge/input_quadrupole_softedge.in:
```diff
@@ -0,0 +1,74 @@
+quad1.type = quadrupole_softedge
+quad1.ds = 1.0
+quad1.gscale = 1.0
+quad1.cos_coefficients =
+          2.0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0
+quad1.sin_coefficients =
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0
```
Wouldn't this work, too?
```suggestion
quad1.cos_coefficients = 2.0
quad1.sin_coefficients = 0.0
```

ax3l commented on examples/quadrupole_softedge/input_quadrupole_softedge.in:
```diff
@@ -0,0 +1,74 @@
+quad2.type = quadrupole_softedge
+quad2.ds = 1.0
+quad2.gscale = -1.0
+quad2.cos_coefficients =
+          2.0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0
+quad2.sin_coefficients =
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0
```
```suggestion
quad2.cos_coefficients = 2.0
quad2.sin_coefficients = 0.0
```

ax3l commented on examples/quadrupole_softedge/run_quadrupole_softedge.py:
```diff
@@ -0,0 +1,223 @@
+quad1 = elements.SoftQuadrupole(
+    ds=1.0,
+    gscale=1.0,
+    cos_coefficients=[
+        2,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+    ],
```
```suggestion
    cos_coefficients=[
        2,
    ],
```

ax3l commented on examples/quadrupole_softedge/run_quadrupole_softedge.py:
```diff
@@ -0,0 +1,223 @@
+    ],
+    sin_coefficients=[
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
```
```suggestion
```

ax3l commented on examples/quadrupole_softedge/run_quadrupole_softedge.py:
```diff
@@ -0,0 +1,223 @@
+    ds=1.0,
+    gscale=-1.0,
+    cos_coefficients=[
+        2,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
```
```suggestion
        2,
```

ax3l commented on examples/quadrupole_softedge/run_quadrupole_softedge.py:
```diff
@@ -0,0 +1,223 @@
+    ],
+    sin_coefficients=[
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
```
```suggestion
```

ax3l commented on src/particles/elements/SoftQuad.H:
```diff
@@ -0,0 +1,479 @@
+
+            // symplectic linear map for a quadrupole is computed using the
+            // Hamiltonian formalism as described in:
+            // https://uspas.fnal.gov/materials/09UNM/ComputationalMethods.pdf.
```
Sometimes safer for links that people might copy:
```suggestion
            // https://uspas.fnal.gov/materials/09UNM/ComputationalMethods.pdf .
```

ax3l reviewed (commented):
This is great, thank you!

I fixed the CUDA issue and added a few inline suggestions.

Feel free to adjust the PR description, it currently reads `Preliminary template only.`?

cemitch99 committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Update input_quadrupole_softedge.in

Reverting back to original coefficient array.

cemitch99 committed:
Update run_quadrupole_softedge.py

Revert to original coefficient array.

cemitch99 committed:
Update input_quadrupole_softedge.in

Revert to original coefficient array.

cemitch99 committed:
Update parameters.rst

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l reviewed (approved):
Thank you, this is great!

I'll investigate separately why the different sized number of coefficients do not work for this and other elements.

ax3l commented:
Ah, I think there is a bug in `ParmParse::queryAdd` and `ParmParse::queryarr` for `std::vector` arguments.
The passed array is only up-sized if needed, but not down-sized if the key exists.
This leads to bugs for long default values and shorter user-provided arrays.
https://github.com/AMReX-Codes/amrex/blob/23.03/Src/Base/AMReX_ParmParse.cpp#L877-L880
